proposals/0025-shader-execution-reordering.md

```diff
@@ -0,0 +1,1460 @@
+visibility rays no longer have to invoke hit shaders in order to access basic
+information about the hit, such as the distance to the closest hit. Finally,
+`HitObject` can be constructed from a `RayQuery`, which enables
+`ReorderThread` and shader table-based closesthit and miss shading to be combined with
+`RayQuery`.
+
+The proposed extension to HLSL should be relatively straightforward to adopt by
+current DXR implementations: `HitObject` merely decouples existing `TraceRay`
+functionality into two distinct stages: the traversal stage and the shading
+stage.
+For SER's `ReorderThread`, the minimal allowed implementation is simply a
+no-op, while implementations that already employ more sophisticated scheduling
+strategies are likely able to reuse existing mechanisms to implement support
+for SER. No DXR runtime changes are necessary, since the proposed extension to
+the programming model is limited to HLSL and DXIL.
+
+## Detailed Design
+
+This section describes the HLSL additions for `HitObject` and `ReorderThread`
+in detail.
+The canonical use of these features involve changing a `TraceRay` call to the
+following sequence:
+
+```C++
+HitObject Hit = HitObject::TraceRay( ..., Payload );
+ReorderThread( Hit );
+HitObject::Invoke( Hit, Payload );
+```
+
+This snippet traces a ray and stores the result of traversal, intersection
+testing and anyhit shading in `Hit`. The call to `ReorderThread` improves
+coherence based on the information inside the `Hit`. Closesthit or miss
+shading is then invoked in a more coherent context.
+
+Note that this is a very basic example. Among other things, it is possible to
+query information about the hit to influence `ReorderThread` with additional
+hints. See [Separation of ReorderThread and HitObject::Invoke](#separation-of-reorderthread-and-hitobjectinvoke)
+for more elaborate examples.
+
+### HitObject HLSL Additions
+
+```C++
+HitObject
+```
+
+The `HitObject` type encapsulates information about a hit or a miss. A
+`HitObject` is constructed using `HitObject::TraceRay`,
+`HitObject::FromRayQuery`, `HitObject::MakeMiss`, or `HitObject::MakeNop`. It
+can be used to invoke closesthit or miss shading using `HitObject::Invoke`,
+and to reorder threads for shading coherence with `ReorderThread`.
+
+The `HitObject` has value semantics, so modifying one `HitObject` will not
+impact any other `HitObject` in the shader. A shader can have any number of
+active `HitObject`s at a time. Each `HitObject` will take up some hardware
+resources to hold information related to the hit or miss, including ray,
+intersection attributes and information about the shader to invoke. The size
+of a `HitObject` is unspecified. As a consequence, `HitObject` cannot be
+stored in structured buffers. Ray payload structs must not have members of
+`HitObject` type. `HitObject` supports assignment (by-value copy) and can be
+passed as arguments to and returned from local functions.
```

Comment:
Description of restriction should mention that the type is `intangible`, and specify more locations that are disallowed.

```suggestion
intersection attributes and information about the shader to invoke. The size
of a `HitObject` is unspecified, and thus considered intangible. As a
consequence, `HitObject` cannot be stored in memory buffers, in groupshared
memory, in ray payloads, or in intersection attributes. `HitObject` supports
assignment (by-value copy) and can be passed as arguments to and returned from
local inlined functions.
```